daemon/logger/loggerutils/logfile.go

```diff
@@ -191,6 +191,16 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		w.currentSize = 0
 		w.notifyRotate.Publish(struct{}{})
 
+		if w.maxFiles <= 1 {
+			go func() {
```

Comment:
Why the goroutine here?

I would think in rotate we can call remove instead of rename when `maxFiles <= 1`

Reply:
Because if we call remove instead of rename, a `fsnotify.Chmod` event is sent and we log stream blocks as before. 

```diff
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index f24dca6260..d6113004de 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -191,16 +191,6 @@ func (w *LogFile) checkCapacityAndRotate() error {
 		w.currentSize = 0
 		w.notifyRotate.Publish(struct{}{})

-		if w.maxFiles <= 1 {
-			go func() {
-				time.Sleep(1 * time.Second)
-				err = os.Remove(fname + ".1")
-				if err != nil && !os.IsNotExist(err) {
-					logrus.Warningf("Failed to remove rotated log file")
-					return
-				}
-			}()
-		}
 		if w.maxFiles <= 1 || !w.compress {
 			w.rotateMu.Unlock()
 			return nil
@@ -216,6 +206,11 @@ func (w *LogFile) checkCapacityAndRotate() error {
 }

 func rotate(name string, maxFiles int, compress bool) error {
+	if maxFiles < 2 {
+		err := os.Remove(name)
+		return err
+	}
+
 	var extension string
 	if compress {
 		extension = ".gz"
@@ -598,6 +593,11 @@ func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan int
 					return err
 				}
 				return nil
+			case fsnotify.Chmod:
+				if err := handleRotate(); err != nil {
+					return err
+				}
+				return nil
 			}
 			return errRetry
 		case err := <-fileWatcher.Errors():
```

I tried this to handle this `fsnotify.Chmod` event. That worked with manual testing but that broke the new test (not in all cases, but the test becomes flaky and sometimes times out on the 2nd log. That's maybe the test is logrotating at a very high frequency because of unrestricted for loop). 

Do you want me to push this diff as a new commit and we try to work on that or should I keep the version with the goroutine ?

Reply:
Hmm, sounds like this issue: https://github.com/fsnotify/fsnotify/issues/194

Reply:
Have you tried by closing the file, then removing, then opening the new file?

Reply:
>Hmm, sounds like this issue: fsnotify/fsnotify#194

Yes

>Have you tried by closing the file, then removing, then opening the new file?

Yes, and in fact it's already done in before the `rotate` call here https://github.com/moby/moby/pull/38489/files/a9e4a4a0c6455c84192f3886e28757d2da27e3a8#diff-0d16783edb4c661112478f7e13a17694R177.